# Project 09. Clustering Grocery Items

In [243]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

# Table of Contents
  1. [Read Data](#read_data)

<a id='read_data'></a>
## Read Data

In [289]:
item_to_id = pd.read_csv('C:/Users/xiaoj/Desktop/CS/collection of data science-take home challenges/Data/9. grocery/item_to_id.csv')
item_to_id.head()

Item_name  Item_id
0           coffee       43
1              tea       23
2            juice       38
3             soda        9
4  sandwich loaves       39

In [290]:
item_to_id.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48 entries, 0 to 47
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Item_name  48 non-null     object
 1   Item_id    48 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 896.0+ bytes


In [308]:
purchase_history = pd.read_csv('C:/Users/xiaoj/Desktop/CS/collection of data science-take home challenges/Data/9. grocery/purchase_history.csv')
purchase_history.head()

user_id                                                 id
0   222087                                              27,26
1  1343649                                            6,47,17
2   404134                       18,12,23,22,27,43,38,20,35,1
3  1110200                                 9,23,2,20,26,47,37
4   224107  31,18,5,13,1,21,48,16,26,2,44,32,20,37,42,35,4...

In [309]:
purchase_history.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39474 entries, 0 to 39473
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  39474 non-null  int64 
 1   id       39474 non-null  object
dtypes: int64(1), object(1)
memory usage: 616.9+ KB


<a id='Data Processing'></a>
## Data Processing

In [310]:
# Built a function to calculate how many times the user bought one item
from collections import Counter
def item_counts(data):
    # transfer the item id from comma-separated list of items to a list
    item_ids = data['id'].str.split(',').sum()
    # to calculate the occurences of all items in a list, there are two menthods: 
    # 1. list comprehension: [[x, data.count(x)[] for x in data]
    # 2. From collections import Counter, Counter(int(id)) for id in item_ids
    return pd.Series(Counter(int(id) for id in item_ids))

In [311]:
user_item_counts = purchase_history.groupby('user_id').apply(item_counts).unstack(fill_value=0)
user_item_counts.head(5)

1   2   3   4   5   6   7   8   9   10  ...  39  40  41  42  43  44  \
user_id                                          ...                           
47        0   1   1   1   0   0   0   0   0   0  ...   0   0   0   0   0   1   
68        0   0   0   0   0   1   0   0   0   1  ...   1   0   0   1   0   0   
113       0   0   1   0   0   0   0   0   1   0  ...   0   0   0   0   1   0   
123       0   0   0   1   0   0   0   0   0   1  ...   0   0   0   0   0   0   
223       1   1   0   0   0   1   0   0   0   0  ...   0   0   1   0   0   0   

         45  46  47  48  
user_id                  
47        1   1   0   0  
68        0   0   0   0  
113       0   1   0   0  
123       0   0   0   0  
223       1   0   0   0  

[5 rows x 48 columns]

In [240]:
# Here is another method to show how to calculate how many times the user bought one item
from collections import Counter
def item_counts_2(data):
    # transfer the item id from comma-separated list of items to a list
    item_ids = data['id'].str.split(',').sum()
    # to calculate the occurences of all items in a list, two menthods: 
    # 1. list comprehension: [[x, data.count(x)[] for x in data]
    # 2. From collections import Counter, Counter(int(id)) for id in item_ids
    return pd.Series(dict((id,item_ids.count(int(id))) for id in item_ids))

In [40]:
# user_item_counts2 = purchase_history.groupby('user_id').apply(item_counts).unstack(fill_value=0)
# user_item_counts2.head(5)

1   2   3   4   5   6   7   8   9   10  ...  39  40  41  42  43  44  \
user_id                                          ...                           
47        0   1   1   1   0   0   0   0   0   0  ...   0   0   0   0   0   1   
68        0   0   0   0   0   1   0   0   0   1  ...   1   0   0   1   0   0   
113       0   0   1   0   0   0   0   0   1   0  ...   0   0   0   0   1   0   
123       0   0   0   1   0   0   0   0   0   1  ...   0   0   0   0   0   0   
223       1   1   0   0   0   1   0   0   0   0  ...   0   0   1   0   0   0   

         45  46  47  48  
user_id                  
47        1   1   0   0  
68        0   0   0   0  
113       0   1   0   0  
123       0   0   0   0  
223       1   0   0   0  

[5 rows x 48 columns]

In [295]:
user_item_counts.shape

(24885, 48)

In [241]:
# This is a test to check what's the difference between w/w/o .sum()(compared with the following[98]
user_buy_items2 = pd.DataFrame(purchase_history.groupby('user_id')['id'])
user_buy_items2

0                                                  1
0           47  32183    3,4,2,13,44,16,24,28,23,45,15,46,25
N...
1           68  35296    23,10,42,39,19,6,17,25
Name: id, dtyp...
2          113  27574    46,33,30,32,43,9,38,21,15,3,25
Name: ...
3          123        38914    4,10,27,18
Name: id, dtype: object
4          223  10297    1,6,35,29,41,45,32
23045            1...
...        ...                                                ...
24880  1499730              1882    45,25
Name: id, dtype: object
24881  1499778  16595    32,2,42,26
22185       14,48,6
Name: ...
24882  1499797  13725    34,37,39,13,44,45,48,11,3,6
Name: id,...
24883  1499817  21544              21,32,42,10,44,17,47,45,8,1...
24884  1499974  27702    34,5,31,11,7,25,8,10,37,13,21,23,39
N...

[24885 rows x 2 columns]

In [242]:
user_buy_items2.loc[0,1]

32183    3,4,2,13,44,16,24,28,23,45,15,46,25
Name: id, dtype: object

In [250]:
user_buy_items3 = pd.DataFrame(purchase_history.groupby('user_id')['id'].sum())
user_buy_items3

id
user_id                                                   
47                     3,4,2,13,44,16,24,28,23,45,15,46,25
68                                  23,10,42,39,19,6,17,25
113                         46,33,30,32,43,9,38,21,15,3,25
123                                             4,10,27,18
223                           1,6,35,29,41,45,3217,2,22,38
...                                                    ...
1499730                                              45,25
1499778                                  32,2,42,2614,48,6
1499797                        34,37,39,13,44,45,48,11,3,6
1499817  21,32,42,10,44,17,47,45,8,16,26,2,10,48,44,38,...
1499974                34,5,31,11,7,25,8,10,37,13,21,23,39

[24885 rows x 1 columns]

In [235]:
user_buy_items3.shape

(24885, 2)

In [199]:
# This is a test to try to find out how to calculate the amount of items for one user 
temp = user_buy_items.loc[47,'id']
print(temp)
len(temp.split(','))

3,4,2,13,44,16,24,28,23,45,15,46,25


13

In [225]:
#I don't know whey this comprehension does not work
user_buy_items.index
(print (len(user_buy_items.loc[i,'id'].split(','))))for i in user_buy_items.index

SyntaxError: invalid syntax (<ipython-input-225-6f7535bc6fc1>, line 3)

In [233]:
def amount(data):
    a = data.loc[:,'id']
    list=[]
    for i in a.index:
        list.append(len(data.loc[i,'id'].split(',')))
    return list
user_buy_items3['amount']=amount(user_buy_items3)
user_buy_items3.sort_values(by='amount', ascending=False)


id  amount
user_id                                                           
269335   2,1,42,11,9,39,37,19,38,2346,14,2,24,4,18,9,6,...      67
367872   17,47,44,25,28,41,2,42,33,11,23,43,1,43,2,2142...      65
397623   20,18,40,32,6,28,41,39,21,48,44,42,3,19,43,9,3...      60
599172   2,30,39,10,37,20,34,18,36,13,1,3233,9,38,6,1,3...      59
377284   15,38,44,43,23,1,19,13,422,12,38,32,20,33,6,45...      59
...                                                    ...     ...
1058759                                                 37       1
1356929                                                 20       1
491450                                                  43       1
490736                                                   2       1
121112                                                  16       1

[24885 rows x 2 columns]

<a id='question1'></a>
## Question 1
### The customer who bought the most items overall in her lifetime

In [313]:
user_item_total = user_item_counts.sum(axis=1)
user_item_total.sort_values(ascending=False).head()

user_id
269335    72
367872    70
599172    64
397623    64
377284    63
dtype: int64

In [314]:
# This is so wired, I DONOT know why and need to check
user_item_total.argmax()

4445

The ID of customer who bought the most items overall in her/his life is 269335, and she/he bought 72 items.

<a id='question2'></a>
## Question 2
For each item, the customer who bought that product the most

In [297]:
user_item_counts.head()

1   2   3   4   5   6   7   8   9   10  ...  39  40  41  42  43  44  \
user_id                                          ...                           
47        0   1   1   1   0   0   0   0   0   0  ...   0   0   0   0   0   1   
68        0   0   0   0   0   1   0   0   0   1  ...   1   0   0   1   0   0   
113       0   0   1   0   0   0   0   0   1   0  ...   0   0   0   0   1   0   
123       0   0   0   1   0   0   0   0   0   1  ...   0   0   0   0   0   0   
223       1   1   0   0   0   1   0   0   0   0  ...   0   0   1   0   0   0   

         45  46  47  48  
user_id                  
47        1   1   0   0  
68        0   0   0   0  
113       0   1   0   0  
123       0   0   0   0  
223       1   0   0   0  

[5 rows x 48 columns]

In [298]:
# to find the custer id who buy the max products
max_user = user_item_counts.apply(lambda a:pd.Series([a.argmax(), a.max()],index=['max_user_ID', 'max_count']))
max_user

1    2     3   4     5     6     7     8     9      10  ...  \
max_user_ID  512  512  2552  92  3605  5555  2926  2493  4445  10238  ...   
max_count      4    5     4   3     3     4     3     3     4      4  ...   

               39   40    41    42     43   44     45     46    47    48  
max_user_ID  9918  634  2192  1303  16510  512  19857  20214  6419  5575  
max_count       5    4     4     4      4    4      5      4     4     3  

[2 rows x 48 columns]

In [299]:
max_user = max_user.transpose()
max_user.index.name = 'Item_ID'
max_user

max_user_ID  max_count
Item_ID                        
1                512          4
2                512          5
3               2552          4
4                 92          3
5               3605          3
6               5555          4
7               2926          3
8               2493          3
9               4445          4
10             10238          4
11              6111          3
12              9261          3
13             10799          4
14              2855          3
15              2356          3
16              1175          3
17              6085          4
18             15193          5
19               512          3
20             14623          4
21             14597          4
22             19892          4
23             15251          5
24              3154          3
25              1094          4
26             16026          4
27             15840          4
28              3400          4
29              6565          4
30               375          2
31              4803          3
32              1771          4
33             21668          3
34              5085          4
35              7526          3
36              4445          4
37               749          4
38              4231          4
39              9918          5
40               634          4
41              2192          4
42              1303          4
43             16510          4
44               512          4
45             19857          5
46             20214          4
47              6419          4
48              5575          3

In [305]:
# Join with item name
#item_to_id = item_to_id.set_index('item_ID')
max_user.join(item_to_id)[['Item_name','max_user_ID','max_count']]

Item_name  max_user_ID  max_count
Item_ID                                           
1                      tea          512          4
2                    juice          512          5
3                     soda         2552          4
4          sandwich loaves           92          3
5             dinner rolls         3605          3
6                tortillas         5555          4
7                   bagels         2926          3
8        canned vegetables         2493          3
9          spaghetti sauce         4445          4
10                 ketchup        10238          4
11                 cheeses         6111          3
12                    eggs         9261          3
13                    milk        10799          4
14                  yogurt         2855          3
15                  butter         2356          3
16                 cereals         1175          3
17                   flour         6085          4
18                   sugar        15193          5
19                   pasta          512          3
20                 waffles        14623          4
21       frozen vegetables        14597          4
22               ice cream        19892          4
23                 poultry        15251          5
24                    beef         3154          3
25                    pork         1094          4
26                 bananas        16026          4
27                 berries        15840          4
28                cherries         3400          4
29              grapefruit         6565          4
30                  apples          375          2
31                broccoli         4803          3
32                 carrots         1771          4
33             cauliflower        21668          3
34               cucumbers         5085          4
35                 lettuce         7526          3
36       laundry detergent         4445          4
37            dishwashing           749          4
38            paper towels         4231          4
39            toilet paper         9918          5
40           aluminum foil          634          4
41           sandwich bags         2192          4
42                 shampoo         1303          4
43                    soap        16510          4
44               hand soap          512          4
45           shaving cream        19857          5
46              baby items        20214          4
47               pet items         6419          4
48                     NaN         5575          3

In [321]:
# This is to show we can also use pd.concat instead of join
final = pd.concat([max_user,item_to_id])
final.set_index('Item_id')

max_user_ID  max_count      Item_name
Item_id                                       
NaN            512.0        4.0            NaN
NaN            512.0        5.0            NaN
NaN           2552.0        4.0            NaN
NaN             92.0        3.0            NaN
NaN           3605.0        3.0            NaN
...              ...        ...            ...
35.0             NaN        NaN           soap
29.0             NaN        NaN      hand soap
19.0             NaN        NaN  shaving cream
4.0              NaN        NaN     baby items
3.0              NaN        NaN      pet items

[96 rows x 3 columns]

### Build the item-item similarity matrix

In [331]:
# To normalize the user_item_counts
from sklearn.preprocessing import normalize
normalized_user_item_counts= normalize(user_item_counts.values, axis=0)
items_similarity = normalized_user_item_counts.T.dot(normalized_user_item_counts)
items_similarity = pd.DataFrame(items_similarity, index=user_item_counts.columns, columns=user_item_counts.columns)
items_similarity.head()

1         2         3         4         5         6         7   \
1  1.000000  0.506895  0.420145  0.296986  0.271132  0.388250  0.271743   
2  0.506895  1.000000  0.466874  0.322744  0.285125  0.468199  0.312200   
3  0.420145  0.466874  1.000000  0.277325  0.224537  0.358326  0.238133   
4  0.296986  0.322744  0.277325  1.000000  0.162860  0.247414  0.166012   
5  0.271132  0.285125  0.224537  0.162860  1.000000  0.233618  0.164699   

         8         9         10  ...        39        40        41        42  \
1  0.335303  0.403690  0.390641  ...  0.388034  0.390286  0.358599  0.393056   
2  0.390521  0.464872  0.527894  ...  0.462968  0.462548  0.409401  0.529100   
3  0.301868  0.362091  0.352597  ...  0.351093  0.368199  0.309078  0.357794   
4  0.216166  0.252662  0.258313  ...  0.245623  0.261272  0.210767  0.244946   
5  0.203245  0.239445  0.234254  ...  0.235543  0.238557  0.211006  0.238466   

         43        44        45        46        47        48  
1  0.395696  0.396766  0.390253  0.394998  0.392164  0.328221  
2  0.464579  0.527325  0.521058  0.462407  0.460257  0.380077  
3  0.351209  0.362522  0.361922  0.354933  0.351832  0.297972  
4  0.253282  0.253119  0.250190  0.253835  0.260541  0.218717  
5  0.235064  0.241835  0.238087  0.238247  0.232387  0.188269  

[5 rows x 48 columns]

In [4]:
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.metrics import silhouette_score